From b86534720d9e25360089fcfbb89d97b81f742a37 Mon Sep 17 00:00:00 2001
From: Rich Bono <rbono@redhat.com>
Date: Thu, 20 Nov 2014 21:23:28 -0500
Subject: [scsi] pm80xx: Fix missing NULL pointer checks and memory leaks

Message-id: <1416518623-7080-8-git-send-email-rbono@redhat.com>
Patchwork-id: 100405
O-Subject: [RHEL7.1 e-stor PATCH V3 07/22] pm80xx : Fix missing NULL pointer checks and memory leaks
Bugzilla: 1110943
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>

From: Suresh Thiagarajan <Suresh.Thiagarajan@pmcs.com>

Checking return value for the memory allocattion and freeing it
while exiting the function

Signed-off-by: Viswas G <Viswas.G@pmcs.com>
Signed-off-by: Suresh Thiagarajan <Suresh.Thiagarajan@pmcs.com>
Reviewed-by: Tomas Henzl <thenzl@redhat.com>
Acked-by: Jack Wang <xjtuwjp@gmail.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit b42939aa075cc45ceb16c18f3b91a36bacbe7813)

Signed-off-by: Rich Bono <rbono@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/pm8001/pm8001_ctl.c b/drivers/scsi/pm8001/pm8001_ctl.c
index 62c884e..a368d77 100644
--- a/drivers/scsi/pm8001/pm8001_ctl.c
+++ b/drivers/scsi/pm8001/pm8001_ctl.c
@@ -395,6 +395,8 @@ static ssize_t pm8001_ctl_bios_version_show(struct device *cdev,
  payload.offset = 0;
  payload.length = 4096;
  payload.func_specific = kzalloc(4096, GFP_KERNEL);
+ if (!payload.func_specific)
+  return -ENOMEM;
  PM8001_CHIP_DISP->get_nvmd_req(pm8001_ha, &payload);
  wait_for_completion(&completion);
  virt_addr = pm8001_ha->memoryMap.region[NVMD].virt_ptr;
@@ -402,6 +404,7 @@ static ssize_t pm8001_ctl_bios_version_show(struct device *cdev,
   bios_index++)
   str += sprintf(str, "%c",
    *((u8 *)((u8 *)virt_addr+bios_index)));
+ kfree(payload.func_specific);
  return str - buf;
 }
 static DEVICE_ATTR(bios_version, S_IRUGO, pm8001_ctl_bios_version_show, NULL);
-- 
1.7.1